File: backend/receipt_service.py
"""Receipt OCR + item classification + anti-fraud (Gemini Vision).

A shop photographs its write-off receipt; one VLM call extracts the merchant,
date, total and line items, normalizes item names and maps each item onto the
platform's four lot categories. Deterministic anti-fraud rules then score the
receipt (stale/future date, duplicates, AI authenticity verdict) before any
lot is created from it.

Synchronous httpx client (same style as ai_service.py) — endpoints calling it
are sync `def`, so FastAPI runs them in its threadpool.
"""
import base64
import hashlib
import json
import logging
import os
from datetime import date, datetime, timedelta, timezone
from typing import Any, Dict, List, Optional

import httpx

GEMINI_API_KEY = os.getenv("GEMINI_API_KEY", "") or os.getenv("GOOGLE_API_KEY", "")
# Vision-capable flash model: receipts are short documents, flash is enough.
OCR_MODEL = os.getenv("OCR_MODEL", os.getenv("AI_MODEL", "gemini-2.5-flash"))

# The platform's lot categories (must match the frontend selects).
LOT_CATEGORIES = ["Выпечка", "Овощи/Фрукты", "Готовая еда", "Молочные продукты"]

# Anti-fraud thresholds.
RECEIPT_MAX_AGE_HOURS = int(os.getenv("RECEIPT_MAX_AGE_HOURS", "48"))
FRAUD_REJECT_THRESHOLD = 0.7   # score >= → receipt rejected outright
FRAUD_FLAG_THRESHOLD = 0.4     # score >= → parsed but flagged for attention

SYSTEM_PROMPT = f"""\
Ты — система распознавания кассовых чеков платформы SaveFood (Казахстан).
На фото — чек списания продуктов из магазина (язык: русский/казахский).

Верни СТРОГО один JSON-объект без пояснений:
{{
  "is_receipt": true|false,
  "merchant": "название магазина или null",
  "receipt_date": "YYYY-MM-DD или null",
  "total": число или null,
  "currency": "KZT" или другая валюта, или null,
  "items": [
    {{
      "raw_name": "позиция как напечатана в чеке",
      "name": "короткое чистое название на русском",
      "category": "одна из: {', '.join(LOT_CATEGORIES)}",
      "quantity": число,
      "unit": "kg" | "pcs" | "l",
      "weight_kg": оценка веса позиции в килограммах (число)
    }}
  ],
  "authenticity": "ok" | "suspicious",
  "authenticity_reason": "почему suspicious, иначе null"
}}

Правила:
- is_receipt=false, если на фото не кассовый/товарный чек.
- category — ближайшая из четырёх; напитки и бакалею относи к "Готовая еда".
- weight_kg: для штучных товаров оцени типичный вес (батон ~0.4, йогурт ~0.3).
- Непродовольственные позиции (пакеты, бытовая химия) пропускай.
- authenticity="suspicious", если чек выглядит отредактированным, нарисованным,
  сфотографированным с экрана, со «скачущими» шрифтами или замазанными полями.
"""


def parse_receipt_image(content: bytes, mime_type: str) -> Optional[Dict[str, Any]]:
    """One Vision call: OCR + normalization + classification + authenticity.

    Returns the parsed dict or None when the AI stage is unavailable/failed
    (no key, network error, unparseable answer).
    """
    if not GEMINI_API_KEY or not content:
        return None
    try:
        with httpx.Client(timeout=60) as client:
            resp = client.post(
                f"https://generativelanguage.googleapis.com/v1beta/models/{OCR_MODEL}:generateContent",
                # Key in a header — never in the query string (see ai_service.py).
                headers={
                    "x-goog-api-key": GEMINI_API_KEY,
                    "content-type": "application/json",
                },
                json={
                    "system_instruction": {"parts": [{"text": SYSTEM_PROMPT}]},
                    "contents": [{
                        "role": "user",
                        "parts": [
                            {"inline_data": {
                                "mime_type": mime_type or "image/jpeg",
                                "data": base64.b64encode(content).decode(),
                            }},
                            {"text": "Распознай этот чек."},
                        ],
                    }],
                    "generationConfig": {
                        "responseMimeType": "application/json",
                        "maxOutputTokens": 4000,
                    },
                },
            )
        if resp.status_code != 200:
            logging.warning("[ocr] Gemini API %s: %s", resp.status_code, resp.text[:200])
            return None
        candidates = resp.json().get("candidates", [])
        if not candidates:
            return None
        parts = candidates[0].get("content", {}).get("parts", [])
        text = "".join(p.get("text", "") for p in parts).strip()
        # responseMimeType=application/json normally yields bare JSON, but be
        # defensive about ```json fences anyway.
        if text.startswith("```"):
            text = text.strip("`")
            if text.startswith("json"):
                text = text[4:]
        parsed = json.loads(text)
        if not isinstance(parsed, dict):
            return None
        return _sanitize(parsed)
    except Exception as e:
        logging.warning("[ocr] receipt parse failed: %s", e)
        return None


def _sanitize(parsed: Dict[str, Any]) -> Dict[str, Any]:
    """Coerce model output into the shapes the rest of the code relies on."""
    items: List[Dict[str, Any]] = []
    for it in parsed.get("items") or []:
        if not isinstance(it, dict):
            continue
        name = (it.get("name") or it.get("raw_name") or "").strip()
        if not name:
            continue
        category = it.get("category")
        if category not in LOT_CATEGORIES:
            category = "Готовая еда"
        try:
            weight = max(0.0, float(it.get("weight_kg") or 0))
        except (TypeError, ValueError):
            weight = 0.0
        items.append({
            "raw_name": (it.get("raw_name") or name).strip(),
            "name": name,
            "category": category,
            "quantity": it.get("quantity"),
            "unit": it.get("unit") if it.get("unit") in ("kg", "pcs", "l") else "pcs",
            "weight_kg": round(weight, 2),
        })
    receipt_date = None
    raw_date = parsed.get("receipt_date")
    if raw_date:
        try:
            receipt_date = date.fromisoformat(str(raw_date)[:10])
        except ValueError:
            receipt_date = None
    total = None
    try:
        if parsed.get("total") is not None:
            total = float(parsed["total"])
    except (TypeError, ValueError):
        total = None
    return {
        "is_receipt": bool(parsed.get("is_receipt")),
        "merchant": (parsed.get("merchant") or None),
        "receipt_date": receipt_date,
        "total": total,
        "currency": parsed.get("currency") or None,
        "items": items,
        "authenticity": "suspicious" if parsed.get("authenticity") == "suspicious" else "ok",
        "authenticity_reason": parsed.get("authenticity_reason") or None,
    }


def sha256_hex(content: bytes) -> str:
    return hashlib.sha256(content).hexdigest()


def fingerprint(parsed: Dict[str, Any]) -> Optional[str]:
    """merchant|date|total — catches the same receipt re-photographed."""
    if not parsed.get("receipt_date") or parsed.get("total") is None:
        return None
    merchant = (parsed.get("merchant") or "").strip().lower()
    return f"{merchant}|{parsed['receipt_date'].isoformat()}|{parsed['total']:.2f}"


def evaluate_fraud(parsed: Dict[str, Any], fingerprint_dupe: bool) -> Dict[str, Any]:
    """Deterministic scoring on top of the AI verdict.

    Cheap rules first (date sanity, duplicates) — they catch most abuse;
    the AI authenticity opinion only adds weight, it never decides alone.
    """
    score = 0.0
    reasons: List[str] = []
    today = datetime.now(timezone.utc).date()

    rdate = parsed.get("receipt_date")
    if rdate is None:
        score += 0.3
        reasons.append("дата чека не распознана")
    elif rdate > today:
        score += 0.8
        reasons.append(f"дата чека в будущем ({rdate.isoformat()})")
    elif rdate < today - timedelta(hours=RECEIPT_MAX_AGE_HOURS):
        score += 0.8
        reasons.append(f"чек старше {RECEIPT_MAX_AGE_HOURS} ч ({rdate.isoformat()})")

    if fingerprint_dupe:
        score += 0.9
        reasons.append("дубликат: чек с тем же магазином, датой и суммой уже загружался")

    if parsed.get("authenticity") == "suspicious":
        score += 0.5
        reason = parsed.get("authenticity_reason") or "признаки редактирования изображения"
        reasons.append(f"ИИ: {reason}")

    score = min(1.0, round(score, 2))
    return {
        "score": score,
        "reasons": reasons,
        "rejected": score >= FRAUD_REJECT_THRESHOLD,
        "flagged": score >= FRAUD_FLAG_THRESHOLD,
    }


def suggest_lots(items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Group parsed items by category into lot drafts the shop can confirm.

    One lot per category (not per line item) — a 15-line receipt must not
    flood the map with 15 tiny lots.
    """
    groups: Dict[str, Dict[str, Any]] = {}
    for it in items:
        g = groups.setdefault(it["category"], {"names": [], "weight": 0.0})
        if it["name"] not in g["names"]:
            g["names"].append(it["name"])
        g["weight"] += it.get("weight_kg") or 0.0
    drafts = []
    for category, g in groups.items():
        # lots.quantity is integer kg; never suggest 0 for a non-empty group.
        quantity = max(1, round(g["weight"]))
        drafts.append({
            "category": category,
            "description": ", ".join(g["names"])[:300],
            "quantity": quantity,
        })
    return drafts

File: backend-java/src/main/java/kz/savefood/receipt/ReceiptService.java
package kz.savefood.receipt;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Duration;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.logging.Logger;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

/**
 * Receipt OCR + item classification + anti-fraud (Gemini Vision), ported from
 * receipt_service.py. One Vision call extracts merchant/date/total/items and maps
 * each item onto the four lot categories; deterministic rules then score the
 * receipt for fraud before any lot is created. Parsed receipts are modelled as
 * {@code Map<String,Object>} to mirror the Python dicts the callers consume.
 */
@Service
public class ReceiptService {

    private static final Logger log = Logger.getLogger(ReceiptService.class.getName());

    public static final List<String> LOT_CATEGORIES =
        List.of("Выпечка", "Овощи/Фрукты", "Готовая еда", "Молочные продукты");

    public static final double FRAUD_REJECT_THRESHOLD = 0.7;
    public static final double FRAUD_FLAG_THRESHOLD = 0.4;

    private static final String SYSTEM_PROMPT = """
        Ты — система распознавания кассовых чеков платформы SaveFood (Казахстан).
        На фото — чек списания продуктов из магазина (язык: русский/казахский).

        Верни СТРОГО один JSON-объект без пояснений:
        {
          "is_receipt": true|false,
          "merchant": "название магазина или null",
          "receipt_date": "YYYY-MM-DD или null",
          "total": число или null,
          "currency": "KZT" или другая валюта, или null,
          "items": [
            {
              "raw_name": "позиция как напечатана в чеке",
              "name": "короткое чистое название на русском",
              "category": "одна из: Выпечка, Овощи/Фрукты, Готовая еда, Молочные продукты",
              "quantity": число,
              "unit": "kg" | "pcs" | "l",
              "weight_kg": оценка веса позиции в килограммах (число)
            }
          ],
          "authenticity": "ok" | "suspicious",
          "authenticity_reason": "почему suspicious, иначе null"
        }

        Правила:
        - is_receipt=false, если на фото не кассовый/товарный чек.
        - category — ближайшая из четырёх; напитки и бакалею относи к "Готовая еда".
        - weight_kg: для штучных товаров оцени типичный вес (батон ~0.4, йогурт ~0.3).
        - Непродовольственные позиции (пакеты, бытовая химия) пропускай.
        - authenticity="suspicious", если чек выглядит отредактированным, нарисованным,
          сфотографированным с экрана, со «скачущими» шрифтами или замазанными полями.
        """;

    private final ObjectMapper mapper = new ObjectMapper();
    private final HttpClient http = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(60)).build();

    private final String apiKey;
    private final String model;
    private final int maxAgeHours;

    public ReceiptService(@Value("${savefood.gemini-api-key:}") String apiKey,
                          @Value("${savefood.ocr-model:gemini-2.5-flash}") String model,
                          @Value("${savefood.receipt-max-age-hours:48}") int maxAgeHours) {
        this.apiKey = apiKey;
        this.model = model;
        this.maxAgeHours = maxAgeHours;
    }

    /**
     * One Vision call: OCR + normalization + classification + authenticity.
     *
     * @return the parsed map, or {@code null} when the AI stage is unavailable
     *         or failed (no key, network error, unparseable answer).
     */
    public Map<String, Object> parseReceiptImage(byte[] content, String mimeType) {
        if (apiKey == null || apiKey.isBlank() || content == null || content.length == 0) {
            return null;
        }
        try {
            Map<String, Object> inlineData = new LinkedHashMap<>();
            inlineData.put("mime_type", mimeType == null || mimeType.isBlank() ? "image/jpeg" : mimeType);
            inlineData.put("data", Base64.getEncoder().encodeToString(content));
            Map<String, Object> body = Map.of(
                "system_instruction", Map.of("parts", List.of(Map.of("text", SYSTEM_PROMPT))),
                "contents", List.of(Map.of(
                    "role", "user",
                    "parts", List.of(
                        Map.of("inline_data", inlineData),
                        Map.of("text", "Распознай этот чек.")))),
                "generationConfig", Map.of(
                    "responseMimeType", "application/json",
                    "maxOutputTokens", 4000));

            HttpRequest req = HttpRequest.newBuilder(URI.create(
                    "https://generativelanguage.googleapis.com/v1beta/models/" + model + ":generateContent"))
                .timeout(Duration.ofSeconds(60))
                .header("x-goog-api-key", apiKey)
                .header("content-type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofByteArray(mapper.writeValueAsBytes(body)))
                .build();
            HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            if (resp.statusCode() != 200) {
                log.warning("[ocr] Gemini API " + resp.statusCode() + ": "
                    + resp.body().substring(0, Math.min(200, resp.body().length())));
                return null;
            }
            JsonNode candidates = mapper.readTree(resp.body()).path("candidates");
            if (!candidates.isArray() || candidates.isEmpty()) {
                return null;
            }
            StringBuilder sb = new StringBuilder();
            for (JsonNode p : candidates.get(0).path("content").path("parts")) {
                sb.append(p.path("text").asText(""));
            }
            String text = sb.toString().strip();
            // responseMimeType=application/json normally yields bare JSON, but be
            // defensive about ```json fences anyway.
            if (text.startsWith("```")) {
                text = text.replaceAll("^`+", "").replaceAll("`+$", "");
                if (text.startsWith("json")) {
                    text = text.substring(4);
                }
            }
            JsonNode parsed = mapper.readTree(text);
            if (!parsed.isObject()) {
                return null;
            }
            return sanitize(parsed);
        } catch (Exception e) {
            log.warning("[ocr] receipt parse failed: " + e);
            return null;
        }
    }

    /** Coerce model output into the shapes the rest of the code relies on. */
    private Map<String, Object> sanitize(JsonNode parsed) {
        List<Map<String, Object>> items = new ArrayList<>();
        for (JsonNode it : parsed.path("items")) {
            if (!it.isObject()) {
                continue;
            }
            String name = textOrEmpty(it, "name");
            if (name.isEmpty()) {
                name = textOrEmpty(it, "raw_name");
            }
            name = name.strip();
            if (name.isEmpty()) {
                continue;
            }
            String category = it.path("category").asText(null);
            if (!LOT_CATEGORIES.contains(category)) {
                category = "Готовая еда";
            }
            double weight = 0.0;
            JsonNode w = it.path("weight_kg");
            if (w.isNumber()) {
                weight = Math.max(0.0, w.asDouble());
            } else if (w.isTextual()) {
                try {
                    weight = Math.max(0.0, Double.parseDouble(w.asText()));
                } catch (NumberFormatException ignored) {
                    weight = 0.0;
                }
            }
            String rawName = textOrEmpty(it, "raw_name");
            if (rawName.isBlank()) {
                rawName = name;
            }
            String unit = it.path("unit").asText(null);
            if (!("kg".equals(unit) || "pcs".equals(unit) || "l".equals(unit))) {
                unit = "pcs";
            }
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("raw_name", rawName.strip());
            item.put("name", name);
            item.put("category", category);
            item.put("quantity", it.has("quantity") && it.get("quantity").isNumber()
                ? it.get("quantity").numberValue() : null);
            item.put("unit", unit);
            item.put("weight_kg", round2(weight));
            items.add(item);
        }

        LocalDate receiptDate = null;
        String rawDate = parsed.path("receipt_date").asText(null);
        if (rawDate != null && !rawDate.isBlank() && !"null".equals(rawDate)) {
            try {
                receiptDate = LocalDate.parse(rawDate.substring(0, Math.min(10, rawDate.length())));
            } catch (RuntimeException ignored) {
                receiptDate = null;
            }
        }
        Double total = null;
        JsonNode totalNode = parsed.path("total");
        if (totalNode.isNumber()) {
            total = totalNode.asDouble();
        } else if (totalNode.isTextual()) {
            try {
                total = Double.parseDouble(totalNode.asText());
            } catch (NumberFormatException ignored) {
                total = null;
            }
        }

        Map<String, Object> out = new LinkedHashMap<>();
        out.put("is_receipt", parsed.path("is_receipt").asBoolean(false));
        out.put("merchant", emptyToNull(parsed.path("merchant").asText(null)));
        out.put("receipt_date", receiptDate);
        out.put("total", total);
        out.put("currency", emptyToNull(parsed.path("currency").asText(null)));
        out.put("items", items);
        out.put("authenticity", "suspicious".equals(parsed.path("authenticity").asText(null))
            ? "suspicious" : "ok");
        out.put("authenticity_reason", emptyToNull(parsed.path("authenticity_reason").asText(null)));
        return out;
    }

    public String sha256Hex(byte[] content) {
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256").digest(content);
            StringBuilder hex = new StringBuilder(digest.length * 2);
            for (byte b : digest) {
                hex.append(Character.forDigit((b >> 4) & 0xF, 16));
                hex.append(Character.forDigit(b & 0xF, 16));
            }
            return hex.toString();
        } catch (Exception e) {
            throw new IllegalStateException("SHA-256 unavailable", e);
        }
    }

    /** merchant|date|total — catches the same receipt re-photographed. */
    @SuppressWarnings("unchecked")
    public String fingerprint(Map<String, Object> parsed) {
        Object date = parsed.get("receipt_date");
        Object total = parsed.get("total");
        if (date == null || total == null) {
            return null;
        }
        String merchant = parsed.get("merchant") == null ? ""
            : ((String) parsed.get("merchant")).strip().toLowerCase(Locale.ROOT);
        return merchant + "|" + ((LocalDate) date) + "|"
            + String.format(Locale.ROOT, "%.2f", ((Number) total).doubleValue());
    }

    /** Deterministic scoring on top of the AI verdict. */
    public Map<String, Object> evaluateFraud(Map<String, Object> parsed, boolean fingerprintDupe) {
        double score = 0.0;
        List<String> reasons = new ArrayList<>();
        LocalDate today = LocalDate.now(ZoneOffset.UTC);

        LocalDate rdate = (LocalDate) parsed.get("receipt_date");
        if (rdate == null) {
            score += 0.3;
            reasons.add("дата чека не распознана");
        } else if (rdate.isAfter(today)) {
            score += 0.8;
            reasons.add("дата чека в будущем (" + rdate + ")");
        } else if (rdate.isBefore(today.minusDays(maxAgeHours / 24))) {
            score += 0.8;
            reasons.add("чек старше " + maxAgeHours + " ч (" + rdate + ")");
        }

        if (fingerprintDupe) {
            score += 0.9;
            reasons.add("дубликат: чек с тем же магазином, датой и суммой уже загружался");
        }

        if ("suspicious".equals(parsed.get("authenticity"))) {
            score += 0.5;
            Object reason = parsed.get("authenticity_reason");
            reasons.add("ИИ: " + (reason == null ? "признаки редактирования изображения" : reason));
        }

        score = Math.min(1.0, round2(score));
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("score", score);
        out.put("reasons", reasons);
        out.put("rejected", score >= FRAUD_REJECT_THRESHOLD);
        out.put("flagged", score >= FRAUD_FLAG_THRESHOLD);
        return out;
    }

    /**
     * Group parsed items by category into lot drafts the shop can confirm — one
     * lot per category (not per line), so a 15-line receipt can't flood the map.
     */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> suggestLots(List<Map<String, Object>> items) {
        Map<String, List<String>> names = new LinkedHashMap<>();
        Map<String, Double> weights = new LinkedHashMap<>();
        for (Map<String, Object> it : items) {
            String category = (String) it.get("category");
            String name = (String) it.get("name");
            List<String> ns = names.computeIfAbsent(category, k -> new ArrayList<>());
            if (!ns.contains(name)) {
                ns.add(name);
            }
            Object w = it.get("weight_kg");
            weights.merge(category, w instanceof Number n ? n.doubleValue() : 0.0, Double::sum);
        }
        List<Map<String, Object>> drafts = new ArrayList<>();
        for (Map.Entry<String, List<String>> e : names.entrySet()) {
            String category = e.getKey();
            // lots.quantity is integer kg; never suggest 0 for a non-empty group.
            int quantity = Math.max(1, (int) Math.round(weights.getOrDefault(category, 0.0)));
            String desc = String.join(", ", e.getValue());
            if (desc.length() > 300) {
                desc = desc.substring(0, 300);
            }
            Map<String, Object> draft = new LinkedHashMap<>();
            draft.put("category", category);
            draft.put("description", desc);
            draft.put("quantity", quantity);
            drafts.add(draft);
        }
        return drafts;
    }

    private static String textOrEmpty(JsonNode node, String field) {
        JsonNode v = node.path(field);
        return v.isTextual() ? v.asText() : "";
    }

    private static String emptyToNull(String s) {
        return s == null || s.isEmpty() || "null".equals(s) ? null : s;
    }

    private static double round2(double x) {
        return BigDecimal.valueOf(x).setScale(2, RoundingMode.HALF_EVEN).doubleValue();
    }
}
